Add "firebase functions:shell" for emulation of all functions types

laurenzlong opened the pull request:
### Description

Adds experimental interactive shell to invoke emulated functions. Shell is a full Node.js repl. Each function name becomes a function that can be called with data and options.

### Sample Commands

`firebase functions:shell`
`firebase functions:shell --port=9000`

**Inside Shell**

For non HTTPS functions:

```
# invoke onCreate function
myFunction('new_data')

# invoke onDelete function
myFunction('old_data')

# invoke onUpdate or onWrite function
myFunction({before: 'old_data', after: 'new_data' })

# mock wildcards in path, for example: if the path was input/{uid}
myFunction('data', {params: {uid: 123}})

# mock auth, for example: to mock unauthenticated user
myFunction('data', {auth: {admin: false}}
```

For HTTPS functions, usage is similar to [request](https://www.npmjs.com/package/request)

```
# invoke
myFunction()
myFunction.get()
myFunction.post()

# invoke at sub-path
myFunction('/path')
myFunction.get('/path')
myFunction.post('/path')

# send POST request with form data
myFunc.post('/path', {form: {foo: 'bar' }})
```

mbleigh commented on lib/callableFunction.js:
```diff
@@ -0,0 +1,81 @@
+var _ = require('lodash');
+var request = require('request');
+
+var _isDatabaseFunc = function(eventTrigger) {
```
It's strange that you're defining a class (something you intend to instantiate via `new` but then not giving it any methods. I'd consider refactoring this a bit so that some of these helper methods are either class or instance methods of CallableFunction.

mbleigh commented on lib/callableFunction.js:
```diff
@@ -0,0 +1,81 @@
+};
+
+var _substituteParams = function(resource, params) {
+  var wildcards = resource.match(new RegExp('{[^/{}]*}', 'g'));
```
I'd take a look at using `String#replace()` with a function ([see docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_function_as_a_parameter)) here -- I think the code would end up being cleaner.

mbleigh commented on lib/callableFunction.js:
```diff
@@ -0,0 +1,81 @@
+      } else if (operationType === 'update' || operationType === 'write') {
+        dataPayload = {
+          data: data.before,
+          delta: data.after
```
This doesn't feel right to me. First, why are we massaging the data into `before` and `after`? Is that part of a planned change to the underlying event structure? Second, `delta` is explicitly **not** the same as the data after, it is the difference between the data before and the data after. I'm worried people will end up with inaccurate local testing if they grow misconceptions based on how this functions vs. how it works live (for instance, `#changed` on delta snapshots will be busted in this impl).

laurenzlong replied:
This is actually a likely breaking change coming to the SDK ("before"/"after" instead of delta)

mbleigh commented on lib/callableFunction.js:
```diff
@@ -0,0 +1,81 @@
+  return 'Successfully invoked function.';
+};
+
+var _requestCallBack = function(err, response, body) {
```
Another thing to evaluate for look and feel (no specific suggestions at this time)

mbleigh commented on lib/serve/functions.js:
```diff
@@ -1,158 +1,14 @@
 module.exports = {
-  start: _startEmulator,
-  stop: _stopEmulator
+  start: function(options) {
```
Since you've abstracted this logic away entirely, it's probably worth finding the call sites of these functions and just swapping them to the new FunctionsEmulator import directly.

laurenzlong replied:
I've made both serve/functions.js and serve/hosting.js modules with "start" and "stop". In the future, when we add more targets for serving, this consistent file structure will make serve/index.js cleaner (similar to how deploy/index.js is set up). That's why I still have this file, instead of just importing FunctionsEmulator directly in serve/index.js

mbleigh commented on lib/functionsEmulator.js:
```diff
@@ -0,0 +1,191 @@
+'use strict';
```
Any meaningful changes to this file I should know about? Haven't taken as close a look at this as the rest of the PR.

laurenzlong replied:
The main things to note are:
- URL for HTTPS functions only gets printed out for firebase serve, not for firebase functions:shell
- firebase serve only serves HTTPS functions, firebase functions:shell emulates all functions

laurenzlong commented on commands/functions-shell.js:
```diff
@@ -0,0 +1,60 @@
+var request = require('request');
+var util = require('util');
+
+var FunctionsEmulator = require('../lib/functionsEmulator');
```
done

laurenzlong commented:
I've made changes according to suggestion, and also:
- renamed command to "experimental:functions:shell"
- changed prompt to fire emoji for Mac/Linux, and "firebase >" otherwise. 

The only thing left is adding helper text.

![screen shot 2017-08-22 at 12 43 28 pm](https://user-images.githubusercontent.com/3990804/29584237-f69f2bbe-8737-11e7-94ae-b8219927bc19.png)
